src/Build/BackEnd/Shared/BuildResult.cs

```diff
@@ -480,6 +480,21 @@ public void AddResultsForTarget(string target, TargetResult result)
             _resultsByTarget[target] = result;
         }
 
+        /// <summary>
+        /// Keep the results for specific targets only.
+        /// </summary>
+        /// <param name="targets">The targets whose results should be kept.</param>
+        public void KeepResultsForSpecificTargets(HashSet<string> targets)
+        {
+            foreach (string target in _resultsByTarget.Keys)
+            {
+                if (!targets.Contains(target))
+                {
+                    _ = _resultsByTarget.TryRemove(target, out _);
```

Comment:
(Unsure) as you're enumerating the keys in this loop, doesn't this throw when you continue to enumerate a mutated collection?

Reply:
It's a safe operation. `_resultsByTarget.Keys` returns a snapshot of the keys in the `ConcurrentDictionary`:
https://github.com/microsoft/referencesource/blob/master/mscorlib/system/collections/Concurrent/ConcurrentDictionary.cs#L2027-L2058

Also confirmed in this MSDN forum:
https://social.msdn.microsoft.com/Forums/vstudio/en-US/64f0c314-e3eb-457a-9dad-14d13cf06d68/updating-values-in-a-concurrentdictionary?forum=parallelextensions

My test `OnlySerializeResultsForSpecifiedTargets` in `CacheSerialization_Tests.cs` goes through this code and didn't raise an exception.

Reply:
Is this more or less efficient than something like:
_resultsByTarget = _resultsByTarget.Where(kvp => targetsToKeep.Contains(kvp.Key)).ToDictionary(kvp => kvp.Key, kvp => kvp.Value);
?

Reply:
Seems to be more efficient. Running `OnlySerializeResultsForSpecifiedTargets` (which goes through this code path) three times with `_resultsByTarget` having 1000 targets yielded an average test run time of 43ms whereas your proposal yielded an average test run time of 50ms.